p2p/discovery/mdns/mdns.go

```diff
@@ -197,6 +198,7 @@ func (s *mdnsService) RegisterNotifee(n Notifee) {
 	s.mutex.Lock()
 	s.notifees = append(s.notifees, n)
 	s.mutex.Unlock()
+	s.resolveOnce.Do(s.startResolver)
```

Comment:
What about just having a separate start function? Honestly, we should stop starting services from within constructors in the first place as it always causes problems like this.

Reply:
That's one option. If changing the API is on the table, the other option would be to pass a `[]Notifee` (or a `...Notifee` or a `WithNotifee(Notifee)`) to the constructor, and remove both `RegisterNotifee` and `UnregisterNotifee`.
Arguably, this is the cleaner API, as a newly registered notifee will have to restart the service, and therefore won't have to wait for cache entries to expire.

Reply:
Passing in a set of notifiees sounds reasonable. I'd still consider a separate start method as being able to fully "construct" a libp2p node before actually _starting_ it would be nice... But not necessary.

My concern with the current patch is that it's kind of implicit and only helps with the first notifiee.

Reply:
Makes sense. #1220 implements this solution.

Reply:
Yeah, I like that one...